### PR TITLE
fix(feature): finite float16 gradients in LoFTR FineMatching std on torch < 2.14

### PR DESCRIPTION
**Pull request by mohansree14**
## What does this pull request do?

`FineMatching.forward` computes the fine-level std as `sqrt(clamp(var, min=1e-10))`. #4229 listed this site under "bound is a positive floor — safe on every version", and that holds in float32 and bfloat16. In float16, though, `1e-10` is `0`, so the floor is the singular point of `sqrt`. On torch older than 2.14, clamp passes its gradient through at the bound, so a zero-variance heatmap (one peaked on a single cell) sends `sqrt'(0) = inf` back to the features as `nan`.

Measured with a peaked heatmap (`M=4, W=5`), gradient of `expec_f[:, 2].sum()` w.r.t. `feat_f1`:

| torch | float16 | bfloat16 | float32 |
|---|---|---|---|
| 2.9.1 (main) | **nan** | finite | finite |
| 2.14.0 (main) | finite | finite | finite |
| 2.9.1 and 2.14.0 (this PR) | finite | finite | finite |

`forward` is not under `no_grad` (only `get_fine_match` is), so `expec_f` carries gradients for the fine-level loss in training.

The fix uses the safe-argument pattern from #4229 and AGENTS.md instead of a floor:

```python
floor = torch.full_like(var, 1e-10)
above = var > floor
safe_var = torch.where(above, var, torch.ones_like(var))
std = torch.sum(torch.where(above, safe_var.sqrt(), floor.sqrt()), -1)
```

At or below the floor the value is still `sqrt(floor)` as represented in the input dtype, with zero gradient, which is what torch 2.14 already gave. The forward output is bit-identical to `main` in float16, bfloat16, float32 and float64 (checked with `torch.equal` on 10k values, including exact zeros and exact `1e-10`).

I also went through the rest of #4229's positive-floor list for float16 underflow:

- `conversions.py` (`axis_angle_to_rotation_matrix`, `1e-12`): affected. Fixed separately in #4613.
- `liegroup/se3.py` (`1e-12`) and `camera/distortion_kannala_brandt.py` (`1e-16`): already guarded with `torch.where` on `main` (#4370 for Kannala-Brandt).
- `aliked/aliked.py` (`1e-8`): not affected. Half inputs are promoted to float32 before `eigh`, so the floor never underflows. Its gradient is `nan` for a covariance with repeated eigenvalues, but that happens in float32 and float64 too. It comes from `eigh`'s backward, not the clamp, and is out of scope here.

A grep for clamp floors from `1e-9` down to `1e-99` in `kornia/` finds no other site.

## Related issue or discussion

Follow-up to #4229 (closed): corrects its classification of `fine_matching.py:83` for float16.

## What did you check?

- New `TestFineMatching::test_convention_std_gradient_is_finite_on_a_peaked_heatmap_4229` in `tests/feature/test_loftr.py`, run with `--dtype=float16,bfloat16,float32,float64`:
  - torch 2.9.1, `main`: float16 fails with a `nan` gradient; the other three pass.
  - torch 2.9.1, this PR: 4 passed.
  - torch 2.14.0: 4 passed on both `main` and this PR. As with the #4007 test, the older torch leg is the one that discriminates.
- `pytest tests/feature/test_loftr.py tests/feature/test_matching.py` (torch 2.14.0): 97 passed, 18 skipped.
- `ruff check` and `ruff format --check` (0.16.7) clean on the changed files.

Not run: full `pre-commit-all`, `ty`, the slow pretrained LoFTR tests, torch 2.5.1, CUDA or MPS. CPU only, Windows.

## How did AI help?

Claude (Opus 5) ran the audit of the #4229 floor list, reproduced the `nan` on torch 2.9.1, wrote the fix and the test, ran the checks above, and drafted this description.

### CODE REVIEW

**Review by ducha-aiki — CHANGES_REQUESTED** [review]
## Review — #4614, head `c4fdbe86`

The production diff is right and I verified it bit-exactly. Both my asks are in the test — one of which is an arithmetic error that currently passes only because it lands inside `atol` with 0.01% of slack.

Merge onto `ecbcad561` is clean.

### The defect reproduces on the floor, and your 2.14 row is correct

I built a torch 2.5.1 / py3.11 scratch venv so the version-bound claim could be executed rather than argued. Peaked heatmap `M=4, W=5`, grad of `expec_f[:,2].sum()` w.r.t. `feat_f1`:

| torch | tree | float16 | bfloat16 | float32 | float64 |
| --- | --- | --- | --- | --- | --- |
| 2.14.0 | main | finite | finite | finite | finite |
| 2.14.0 | PR | finite | finite | finite | finite |
| 2.5.1 | main | **800/800 nan** | finite | finite | finite |
| 2.5.1 | PR | finite | finite | finite | finite |

So your table's 2.14 row is confirmed — main is already finite there — and I reproduced the defect on the **2.5.1 floor** (you measured 2.9.1). `std` is `[0,0,0,0]` in fp16 on every leg, so only the gradient moves.

**Forward is bit-exact.** `torch.equal(old_std, new_std)` is True for all four dtypes over 10,750 `var` values (10,000 random in [-2,2], exact `0`, `1e-10`, `-1e-10`, `1e-11`, `1e-9`, `logspace(-20,2,500)`), on both torch versions. End-to-end on seeded random `M=64, W=5, C=16`: `expec_f` **and** `feat_f1.grad` are `torch.equal` between main and this branch for all four dtypes on 2.14.

On 2.5.1 the only gradient changes are fp16 `var ≤ 0` (`±inf → 0`) and `var == 1e-10` exactly in bf16/fp32/fp64 (`49920` / `50000` / `49999.99…` → `0`) — i.e. the fix makes the old torch agree with what 2.14 already returns. Nothing else moves.

**The test discriminates.** Your test file against main's production code on torch 2.5.1: `1 failed, 3 passed, 22 skipped`, the failure being `[cpu-float16]` with an all-nan `feat_f1.grad`.

---

## [P2] The value assertion is arithmetically wrong

`std` is `torch.sum(..., -1)` over the two coordinate axes, so the floored value is `2*sqrt(floor)`, not `sqrt(floor)`. The test asserts `assert_close(std, torch.full_like(std, 1e-10).sqrt())`.

Measured: float32 actual `2.000000e-05` vs asserted `1.000000e-05` — diff `1.000000e-05` against a tolerance of `atol 1e-5 + rtol 1e-4 * 1e-5 = 1.000100e-05`. It passes with **1e-9 of slack, 0.01%**. float64: diff `1.000000e-05` vs tolerance `1.000010e-05`, 0.001%. And `_DTYPE_PRECISIONS`' float32 entry carries a `# TODO: Update to ~1.2e-7` — the day that lands, this test goes red on *correct* code.

Today's consequence: the same assertion also accepts `std == 0`, `std == 1e-10` (sqrt dropped) and a `1e-12` floor. I fed each candidate to `testing.base.assert_close` to confirm.

Validated fix — unmutated `4 passed` on both torches, and it kills mutations 2 and 5 below:

```python
self.assert_close(std, torch.full_like(std, 1e-10).sqrt() * 2, rtol=0.0, atol=0.0)
```

(float16 is unkillable there either way, since `sqrt(1e-10) == 0` on both sides.)

## [P2] Nothing pins the selected (`var > floor`) branch — and nothing else in the repo does either

Mutating `above = var > floor` to `torch.zeros_like(var, dtype=torch.bool)` — floor branch always taken, which corrupts `std` for *every real heatmap* — leaves `TestFineMatching` at `4 passed` on both torches, **and** leaves the full `pytest tests/feature/test_loftr.py --runslow` at `7 passed, 1 skipped` with the real `loftr_indoor.ckpt` / `loftr_outdoor.ckpt` weights (59.8 s control vs 31.5 s mutated).

The reason: no existing test ever reads `expec_f`/`std` — the pretrained tests compare `keypoints0/1`, which come from `coords_normalized` through `get_fine_match`. And every other test in `test_loftr.py` is `@pytest.mark.slow` while `runslow` is never set in `pr_test_cpu.yml`, so on PR CI **your new test is the only `test_loftr.py` coverage that runs.**

Validated fix: add a second case with a uniform heatmap (`feat_f0 = feat_f1 = zeros`, so `var = mean(g²) = 0.5` per axis) asserting `std == 2*sqrt(0.5)`. Unmutated `8 passed`; under the always-False mutation, `4 failed` on all four dtypes.

## [P3] `test_convention_*_4229` is a regression pin, not a convention record

There is no `test_wart_*_4229` on main to invert — `grep -rnE "#4229|issues/4229" kornia/ tests/` on the merge tree returns only prose about *other* sites — and #4229 is an audit issue, not a documented-defect issue. `test_regression_*` or plain `test_std_gradient_is_finite_*_4229` would read more truthfully. Nothing owed under the AGENTS.md grep rule.

## [P3] The changelog fragment understates the blast radius

It says "when a heatmap has zero variance", which reads as a degenerate corner. Measured on 200 batches of 256 random `M×25×16` float16 features: `var` equals the underflowed bound on **10.21%** of coordinate entries, poisoning **6933/51200 = 13.5%** of match rows — and since `std` is summed into the loss, one nan makes the *entire* `feat_f1.grad` nan (I observed 800/800). float32 hit the bound 0/102400 times and bfloat16 1/102400, where the bound is a normal number so the gradient is a finite 49920 — your table is right to mark those finite. Wording only.

---

## Verified as correct

- **The discriminating CI leg exists.** `pr_test_cpu.yml` runs the half-precision job at `pytorch-version: ["2.9.1"]` with `pytorch-dtype: ["float16","bfloat16"]`, so this pin will actually fire on a regression. No `half_precision_xfails/` or `known_failure_xfails/` entry needs adding or removing (grep for loftr/FineMatching in both: empty), and the new test passes on 2.5.1 and 2.9.1 with the fix.
- **`expec_f` really does carry gradients** — executed, not read. `forward` (`fine_matching.py:36`) has no `no_grad`; the decorator is on `get_fine_match` at :97; `loftr.py:201` calls `self.fine_matching(...)` outside any `no_grad`. The probe shows a real `feat_f1.grad` (all-zero, then all-nan on the broken leg).
- **The idiom matches the nearby canonical pattern** in `kornia/geometry/liegroup/se3.py:185-188` and `:217-218`.
- **Your pattern audit is accurate.** No other `sqrt(clamp(...))` in `kornia/feature/loftr/`. Repo-wide the only surviving small-floor sites are `conversions.py:444` (`1e-12`, the subject of sibling PR #4613) and `aliked.py:171` (`clamp(min=1e-8).sqrt()`) — and I confirmed `aliked.py:168-169` promotes float16/bfloat16 `cov` to float32 *before* the clamp, so `1e-8` cannot underflow there.
- **`torch.compile(fullgraph=True)`**: forward and backward identical to the old expression under inductor. The one fp16 eager-vs-compiled delta (`15.8125` vs `15.8046875`) is byte-identical on main, so it is a pre-existing inductor rounding artifact, not introduced here.
- `tests/feature/test_matching.py tests/feature/test_loftr.py` on 2.14: 97 passed, 18 skipped — matches your body. New test on MPS float32: 1 passed. Pinned `ruff 0.16.7` check and format clean, `ty` clean, nothing over 120. `changelog.d/4614.fixed.md` correctly named. No new public names.

## Mutation testing

Control: `4 passed` on both torch 2.14.0 and 2.5.1.

| mutation | result |
| --- | --- |
| `var > floor` → `var >= floor` | **killed** — 1 failed |
| `floor.sqrt()` → `floor` | **SURVIVED** → P2 #1 |
| `safe_var = where(above, var, ones)` → `safe_var = var` | **killed** — 3 failed |
| `where(above, safe_var.sqrt(), floor.sqrt())` → `safe_var.sqrt()` | **killed** — 4 failed |
| floor `1e-10` → `1e-12` | **SURVIVED** → P2 #1 |
| floor `1e-10` → `1e-4` | **killed** — 4 failed |
| safe argument `ones_like` → `zeros_like` | survived — **equivalent mutant**: the inner `where` masks `sqrt`'s `0/0` backward elementwise exactly where `above` is False. Not a weakness. |
| `above` → always-False | **SURVIVED**, including `--runslow` with real weights → P2 #2 |
| `above` → always-True | **killed** — 3 failed |

With both proposed test changes applied, mutations 2, 5 and 8 are all killed.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Review by ducha-aiki — APPROVED** [review]
Re-reviewed the latest delta. The expected value now includes both coordinate axes, the selected non-floor branch is covered, and the fragment accurately describes the trigger. Current CI is green and I found no unresolved findings.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-5).
